switchboard/base.py

```diff
@@ -6,77 +6,91 @@
 :license: Apache License 2.0, see LICENSE for more details.
 """
 
-import time
-import logging
 import threading
 
-from .models import MongoModel
-from .signals import request_finished
-from .settings import settings
 
-log = logging.getLogger(__name__)
+class ModelDict(threading.local):
+    """
+    Dictionary-style access to model data.
+
+    If ``auto_create=True`` accessing modeldict[key] when key does not exist
+    will attempt to create it in the datastore.
 
+    Functions in two different ways, depending on the constructor:
+
+        # Given a model that has a attribute named ``foo`` where the value
+        # is "bar":
 
-NoValue = object()
+        mydict = ModelDict(Model, value='foo')
+        mydict['test']
+        >>> 'bar' #doctest: +SKIP
 
+    If you want to use another key besides ``_id``, you may specify that in the
+    constructor.
 
-class CachedDict(threading.local):
-    def __init__(self, timeout=30):
-        """
-        Not guaranteed to be called with expected c'tor args of
-        all usages (due to usage of threading.local)
-        """
-        cls_name = type(self).__name__
+        mydict = ModelDict(Model, key='foo', value='id')
+        mydict['bar']
+        >>> 'test' #doctest: +SKIP
 
-        self._cache = None
-        self._last_updated = None
-        self.timeout = timeout
-        self.cache = settings.SWITCHBOARD_CACHE
-        self.cache_key = cls_name
-        self.last_updated_cache_key = '%s.last_updated' % (cls_name,)
+    """
+    def __init__(self, model, key='_id', value=None,
+                 auto_create=False, *args, **kwargs):
+        assert value is not None
+
+        super(ModelDict, self).__init__(*args, **kwargs)
+
+        self.key = key
+        self.value = value
+
+        self.model = model
+        self.auto_create = auto_create
 
     def __getitem__(self, key):
-        self._populate()
-        try:
-            value = self._cache[key]
-        except KeyError:
-            value = self.get_default(key)
-            if value is NoValue:
-                raise
-        except (TypeError, AttributeError):
-            log.exception('Unable to access the local cache')
-            value = self.get_default(key)
-            if value is NoValue:
-                raise KeyError(key)
-        return value
+        if self.auto_create:
+            result = self.model.get_or_create(key)[0]
+        else:
+            result = self.model.get(key)
+        if result is None:
+            raise KeyError(key)
+        return result
+
+    def __setitem__(self, key, value):
+        if isinstance(value, self.model):
+            nested_value = getattr(value, self.value)
+        else:
+            nested_value = value
+
+        defaults = {self.value: nested_value}
+        instance, created = self.model.get_or_create(key, defaults=defaults)
+
+        # Ensure we're updating the value in the datastore if it changes.
+        if getattr(instance, self.value) != nested_value:
```

Comment:
Again, why we can be sure that a filed named self.value is in instance? Please add comments.